## Get the sequences form the .bed files
```{bash}
%%bash

for bedfile in /storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass/*.bed
do
    cell_name=$(basename $bedfile .bed)
    bedtools getfasta -fi /storage1/fs1/yeli/Active/yeli/genome/hg38/hg38.fa -bed $bedfile -fo ${cell_name}_sequences.fasta
done
```

```{python}
# Load packages
import csv
import os
from Bio import SeqIO
```

## Use whole dataset
```{python}
# File path
bed_dir = '/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass'
fasta_dir = '/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass'
output_csv = '/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/nt_data/subclass/sampled_humanbrain_cCREs_subclass.csv'

bedfiles=["/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass/ASCT.bed",
         "/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass/OPC.bed",
         "/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass/OGC.bed",
         "/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass/MGC.bed",
         "/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass/ITL23.bed",
         "/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass/VIP.bed"]
# Create .cvs file
with open(output_csv, mode='w', newline='') as csvfile:
    csv_writer = csv.writer(csvfile)
    csv_writer.writerow(['cell', 'chr', 'start', 'end', 'seq'])

    # Read .bed file folder
    for bedfile in bedfiles:
        if bedfile.endswith('.bed'):
            cell_name = bedfile.replace('.bed', '')
            fasta_file = os.path.join(fasta_dir, f'{cell_name}_sequences.fasta')

            # Read .bed files
            bed_filepath = os.path.join(bed_dir, bedfile)
            with open(bed_filepath, 'r') as bed:
                for bed_line, fasta_record in zip(bed, SeqIO.parse(fasta_file, "fasta")):
                    # extract chr, start, end
                    bed_fields = bed_line.strip().split()
                    chr_name, start, end = bed_fields[0], bed_fields[1], bed_fields[2]
                    seq = str(fasta_record.seq)
                    
                    # Write the .csv dataset
                    csv_writer.writerow([cell_name, chr_name, start, end, seq])

print(f"CSV file saved at {output_csv}")
```

## Use randomely sampled datasets
```{bash}
%%bash

# Define input, output and reference genome
INPUT_DIR="/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass"
OUTPUT_DIR="/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/sampled_subclass"
GENOME_FA="/storage1/fs1/yeli/Active/yeli/genome/hg38/hg38.fa"  # path to reference genome

# Create output_dir
mkdir -p $OUTPUT_DIR


for BED_FILE in ASCT.bed OPC.bed OGC.bed MGC.bed ITL23.bed VIP.bed; do
    BASE_NAME=$(basename $BED_FILE .bed)

    # Randomly sample 500 DNA sequences
    SORTED_FILE="$OUTPUT_DIR/${BASE_NAME}_sorted_2k.bed"
    shuf "$INPUT_DIR/$BED_FILE" | head -n 2000 | sort -k1,1 -k2,2n > $SORTED_FILE

    # Get fastq
    FASTA_FILE="$OUTPUT_DIR/${BASE_NAME}_2k.fasta"
    bedtools getfasta -fi $GENOME_FA -bed $SORTED_FILE -fo $FASTA_FILE

    echo "Generated $FASTA_FILE from $BED_FILE"
done
```

```{python}
import csv
import os
from Bio import SeqIO
```

```{python}
import os
import csv
import subprocess
from Bio import SeqIO

# File path
bed_dir = '/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/subclass'
fasta_dir = '/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/raw_data/sampled_subclass'
output_csv = '/storage1/fs1/yeli/Active/l.ronghan/data/humanbrain_cCREs/nt_data/subclass/sampled_2k_humanbrain_cCREs_subclass.csv'
genome_fa = '/storage1/fs1/yeli/Active/yeli/genome/hg38/hg38.fa'

bedfiles = [
    "ASCT.bed",
    "OPC.bed",
    "OGC.bed",
    "MGC.bed",
    "ITL23.bed",
    "VIP.bed"
]

# Create .csv file
with open(output_csv, mode='w', newline='') as csvfile:
    csv_writer = csv.writer(csvfile)
    csv_writer.writerow(['cell', 'chr', 'start', 'end', 'seq'])

    # Process each .bed file
    for bedfile in bedfiles:
        cell_name = bedfile.replace('.bed', '')  # Get cell name without .bed extension
        bed_filepath = os.path.join(bed_dir, bedfile)
        sorted_bed_filepath = os.path.join(fasta_dir, f'{cell_name}_sorted_2k.bed')
        fasta_file = os.path.join(fasta_dir, f'{cell_name}_sorted_2k.fasta')

        # Step 1: Randomly shuffle, select 500 sequences, and sort by chromosome and start position
        sort_command = f"shuf {bed_filepath} | head -n 2000 | sort -k1,1 -k2,2n > {sorted_bed_filepath}"
        subprocess.run(sort_command, shell=True, check=True)

        # Step 2: Use bedtools to get fasta sequences
        bedtools_command = f"bedtools getfasta -fi {genome_fa} -bed {sorted_bed_filepath} -fo {fasta_file}"
        subprocess.run(bedtools_command, shell=True, check=True)

        # Step 3: Read sorted .bed file and corresponding .fasta file
        with open(sorted_bed_filepath, 'r') as bed:
            for bed_line, fasta_record in zip(bed, SeqIO.parse(fasta_file, "fasta")):
                # Extract chr, start, end from bed file
                bed_fields = bed_line.strip().split()
                chr_name, start, end = bed_fields[0], bed_fields[1], bed_fields[2]
                seq = str(fasta_record.seq)

                # Write to CSV file
                csv_writer.writerow([cell_name, chr_name, start, end, seq])

print(f"CSV file saved at {output_csv}")
```
